(WIP) feat(cc-icon): migrate components to new icon system

roberttran-cc opened the pull request:
# Migrating components to new icon system

Here is a document to help migrating from the previous icon system to the new one (when possible).

You'll find below 4 different lists. Each one details, for each SVG file icon, in which component(s) it's used and which Remix Icon could replace it:
- the 1st one proposes Remix Icons that are (almost) identical to the original one,
- the 2nd one proposes Remix Icons that are more or less close to the original one,
- the 3rd one lists icons that have no close match among Remix Icons,
- the 4th one lists icons that have originally multiple colors and that can't be replaced.

Feel free to propose other Remix Icon!

Note: I guess decorative icons in stories can be replaced even though there is no close match.

## How to migrate

When searching for a Remix Icon, it's simpler to browse on the [official site](https://remixicon.com/).

You can check the migration that have been done for `cc-ssh-key-list`: [`e12dd10` (#570)](https://github.com/CleverCloud/clever-components/pull/570/commits/e12dd10cce4ee926346d6e22a306d4e896b0041a)

I guess the way to go is one commit per component:
- when you're done, strike out the migrated component from the list,
- remember that one component can contain multiple icons,
- do not forget to remove the SVG file if you remove its last usage: strike out the whole line if that's the case,
- if you want to/have to keep the legacy icon, you can update the SVG file to benefit from the cc-icon properties (remove hard-coded colors, fix sizes, [optimize](https://jakearchibald.github.io/svgomg/)).

Feel free to share before/after screenshots if you have doubts.

## List 1: identical matches

| Icon file                  | Remix icon proposal           | Impacted components and stories                                                                  |
|----------------------------|-------------------------------|--------------------------------------------------------------------------------------------------|
| `alert-fill.svg`           | `iconRemixAlertFill`          | cc-notice.js<br/>cc-toaster.stories.js                                                           |
| `alert-line.svg`           | `iconRemixAlertLine`          | cc-toast.js                                                                                      |
| `checkbox-circle-fill.svg` | `iconRemixCheckboxCircleFill` | cc-email-list.js<br/>cc-notice.js<br/>cc-toaster.stories.js                                      |
| `checkbox-circle-line.svg` | `iconRemixCheckboxCircleLine` | cc-toast.js                                                                                      |
| `close-gray.svg`           | `iconRemixCloseLine`          | cc-notice.js<br/>cc-toast.js                                                                     |
| `close.svg`                | `iconRemixCloseLine`          | cc-addon-backups.js<br/>cc-button.stories.js<br/>cc-tile-requests.js<br/>cc-tile-status-codes.js |
| `cpu.svg`                  | `iconRemixCpuLine`            | cc-addon-features.js                                                                             |
| `disk.svg`                 | `iconRemixDatabase_2Fill`     | cc-addon-features.js                                                                             |
| `down.svg`                 | `iconRemixArrowDownSFill`     | cc-block.js<br/>cc-pricing-product-consumption.js<br/>cc-pricing-table.js                        |
| `error.svg`                | `iconRemixErrorWarningFill`   | cc-orga-member-card.js<br/>cc-badge.stories.js                                                   |
| `file.svg`                 | `iconRemixFileTextLine`       | cc-invoice.js<br/>cc-invoice-table.js                                                            |
| `information-fill.svg`     | `iconRemixInformationFill`    | cc-notice.js<br/>cc-toaster.stories.js                                                           |
| `information-line.svg`     | `iconRemixInformationLine`    | cc-toast.js                                                                                      |
| `pen.svg`                  | `iconRemixEditFill`           | cc-orga-member-card.js                                                                           |
| `phone.svg`                | `iconRemixPhoneFill`          | cc-header-orga.js                                                                                |
| `profile.svg`              | `iconRemixAccountCircleFill`  | cc-orga-member-card.js                                                                           |
| `spam-2-fill.svg`          | `iconRemixSpam_2Fill`         | cc-email-list.js<br/>cc-notice.js<br/>cc-toaster.stories.js                                      |
| `spam-2-line.svg`          | `iconRemixSpam_2Line`         | cc-toast.js                                                                                      |
| `stopped.svg`              | `iconRemixStopFill`           | cc-header-app.js                                                                                 |
| `storage.svg`              | `iconRemixDatabase_2Fill`     | cc-pricing-product-consumption.js                                                                |
| `sum.svg`                  | `iconRemixFunctions`          | cc-pricing-product-consumption.js                                                                |
| `up.svg`                   | `iconRemixArrowUpSFill`       | cc-block.js<br/>cc-pricing-product-consumption.js<br/>cc-pricing-table.js                        |
| `warning.svg`              | `iconRemixAlertFill`          | cc-error.js<br/>cc-badge.stories.js<br/>cc-button.stories.js                                     |

## List 2: close matches

| Icon file                      | Remix icon proposal          | Impacted components and stories                                     |
|--------------------------------|------------------------------|---------------------------------------------------------------------|
| `backup.svg`                   | `iconRemixHistoryLine`       | cc-addon-backups.js                                                 |
| `clipboard.svg`                | `iconRemixClipboardLine`     | cc-input-text.js                                                    |
| `decrement.svg`                | `iconRemixSubtractLine`      | cc-input-number.js                                                  |
| `delete-neutral.svg`           | `iconRemixCloseLine`         | cc-orga-member-card.js                                              |
| `delete.svg`                   | `iconRemixCloseLine`         | cc-pricing-estimation.js                                            |
| `encryption-at-rest.svg`       | `iconRemixShieldKeyholeFill` | cc-addon-encryption-at-rest-option.js (*)                           |
| `eye-closed.svg`               | `iconRemixEyeOffFill`        | cc-input-text.js                                                    |
| `eye-open.svg`                 | `iconRemixEyeFill`           | cc-input-text.js                                                    |
| `increment.svg`                | `iconRemixAddLine`           | cc-input-number.js                                                  |
| `mail-line.svg`                | `iconRemixMailLine`          | cc-email-list.js                                                    |
| `mail-star-line.svg`           | `iconRemixMailStarLine`      | cc-email-list.js                                                    |
| `plus.svg`                     | `iconRemixAddLine`           | cc-pricing-table.js<br/>cc-pricing-product-consumption.js           |
| `restarting-with-downtime.svg` | `iconRemixPlayLine`          | cc-header-app.js                                                    |
| `running.svg`                  | `iconRemixPlayFill`          | cc-header-app.js<br/>cc-tile-instances.js                           |
| `starting.svg`                 | `iconRemixPlayLine`          | cc-header-app.js<br/>cc-tile-instances.js                           |
| `tick-blue.svg`                | `iconRemixCheckLine`         | cc-orga-member-card.js                                              |
| `tick.svg`                     | `iconRemixCheckLine`         | cc-input-text.js<br/>cc-orga-member-card.js<br/>cc-badge.stories.js |
| `trash-red.svg`                | `iconRemixDeleteBin_5Fill`   | cc-email-list.js<br/>cc-orga-member-card.js                         |
| `unknown.svg`                  | `iconRemixMoreFill`          | cc-header-app.js                                                    |
| `user.svg`                     | `iconRemixUser_3Fill`        | cc-pricing-product-consumption.js                                   |

## List 3: no match

| Icon file             | Remix icon proposal | Impacted components and stories                                                                                                                                                                   |
|-----------------------|---------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `arrows-left.svg`     | `n/a`               | cc-pricing-product-consumption.js                                                                                                                                                                 |
| `arrows-right.svg`    | `n/a`               | cc-pricing-product-consumption.js                                                                                                                                                                 |
| `badge-white.svg`     | `n/a`               | cc-header-orga.js<br/>cc-badge.stories.js                                                                                                                                                         |
| `git.svg`             | `n/a`               | cc-header-app.js                                                                                                                                                                                  |
| `info.svg`            | `n/a`               | cc-button.stories.js<br/>cc-badge.stories.js<br/>cc-elasticsearch-info.js<br/>cc-grafana-info.js<br/>cc-jenkins-info.js<br/>cc-matomo-info.js<br/>cc-tile-requests.js<br/>cc-tile-status-codes.js |
| `notice.svg`          | `n/a`               | cc-error.js                                                                                                                                                                                       |
| `ram.svg`             | `n/a`               | cc-addon-features.js                                                                                                                                                                              |
| `redirection-off.svg` | `n/a`               | cc-tcp-redirection.js                                                                                                                                                                             |
| `redirection-on.svg`  | `n/a`               | cc-tcp-redirection.js                                                                                                                                                                             |
| `start-failed.svg`    | `n/a`               | cc-header-app.js                                                                                                                                                                                  |

## List 4: multiple colors

| Icon file            | Related component  |
|----------------------|--------------------|
| `restart-failed.svg` | `cc-header-app.js` |
| `restarting.svg`     | `cc-header-app.js` |

florian-sanders-cc commented:
:wave: First off @roberttran-cc thanks for this awesome work.

It really makes the migration process easier :sunglasses: 

I was migrating `cc-orga-member-card` and I'm not sure about the Remix equivalents for close and validate button images.
Before:
![image](https://user-images.githubusercontent.com/100240294/211589325-452771f5-ee78-4e77-aa02-bc9dffe8e86a.png)
After:
![image](https://user-images.githubusercontent.com/100240294/211589185-0f5a2a39-d86e-469a-80ec-6762908f361e.png)

The remix ones look smaller and thinner but the current ones might actually be too bold?
WDYT? :smile:

florian-sanders-cc commented:
# Which CSS custom props to use

We discussed the following issue during the team meeting (2023-02-13).

We decided to use CSS custom props related to texts (`--cc-color-text-default`, `--cc-color-text-primary`...) when we need to set a color for an icon.
![image](https://user-images.githubusercontent.com/100240294/218531879-b07a97a2-96b9-478f-ab28-2366d254a766.png)

We cannot do this for icons inside buttons used in `cc-input-*` because we would need a color too light to be used for texts (> 3:1 < 4.5:1 of contrast ratio) so we use a CSS custom prop at component level with a hard coded default value.

We'll open an RFC issue to discuss whether we should introduce icon semantic tokens.

roberttran-cc reviewed (commented):
Thanks for the awesome contributions!

I have some minor feedback but I think I will take care of them myself.
There are also some tasks left: I edited the top message accordingly.

Nevertheless, we're getting really close, thanks again!

roberttran-cc commented on src/components/cc-tcp-redirection/cc-tcp-redirection.js:
```diff
@@ -103,7 +99,11 @@ export class CcTcpRedirection extends LitElement {
+              isRedirectionDefined
+                ? html`<cc-icon .icon=${iconRedirectionOn} class="on"></cc-icon>`
+                : html`<cc-icon .icon=${iconRedirectionOff} class="off"></cc-icon>`
+            }
```
Actually, don't mind this suggestion, it's actually slightly more complex than I thought. :see_no_evil:

hsablonniere commented:
This is impressive work :clap: 

1. I think we should merge this after the pricing components
2. `cc-addon-encryption-at-rest-option` template still uses the `new URL(href, import.meta.url)` technique
3. I also see `new URL(href, import.meta.url)` usage in some stories, not sure it's a problem (see 4)
4. I would be nice if we could completely remove the `@web/rollup-plugin-import-meta-assets` plugin from the rollup builds

pdesoyres-cc commented:
I validate all the new icons. Thank you for making the review straightforward.

florian-sanders-cc reviewed (changes requested):
@roberttran-cc thanks a lot for this wonderful work of coordination, documentation and reorganisation + code fixing (:sweat_smile:)

A few global notes but we're almost there I think:

- should we move `IconModel` interface to `common.types.d.ts` since many component rely on it? 
- I think we should rollback changes made to `cc-pricing-*` components. These changes are only bringing conflict with the `cc-pricing` PR that should be merged soon.
- About the `cc-input-text` eye icon, I think we could try with the not filled icon to make it a litle bit more subtle?
- `cc-invoice-table` (responsive): the icons are too small and it's actually my bad. Need to `flex: 0 0 auto` + `size="lg"`.
- `cc-header-app`: As discussed during the meeting, we should probably rollback (apparently it's actually my bad here too).
- same for `cc-tile-instances`.
- `cc-tile-metrics`: something is off with the colors of the info icon.
- `cc-email-list`: I remember we changed the remix icon because we wanted it to be more rectangular. I think I'd stick with the more rectangular one that we have in prod (but maybe that's just me so let's see what others think of course :wink:)

roberttran-cc commented on src/components/cc-pricing-product-consumption/cc-pricing-product-consumption.js:
```diff
@@ -318,7 +321,7 @@ export class CcPricingProductConsumption extends withResizeObserver(LitElement)
         <div class="section">
           <div class="section-header">
-            <cc-img class="section-icon" src=${sumSvg}></cc-img>
+            <cc-icon class="section-icon" .icon=${iconSum}></cc-icon>
```
Actually, this will be fix within #716!

roberttran-cc commented:
Thanks for your feedback! :heart_decoration: 

I pushed some fixes accordingly, updated the initial post to reflect those changes: this PR is up for review again! :rocket: 

For @florian-sanders-cc & @pdesoyres-cc:
- there are some single named commits for some of your feedback: you can review them independently
- exception made of `cc-pricing-*` whose commits were roll-backed so there's nothing there to see
- exception made of `cc-header-app` & `cc-tile-instances` for which I fixed up their initial commit (for better readability): you should check them completely
- exception made of the import order feedback that will probably be dealt in another issue (we'll talk about that next Monday Meeting)

For the rest of you, statement from the initial post still stands: side-by-side graphic comparison and only then check the updates in the code!

florian-sanders-cc reviewed (changes requested):
Gg @roberttran-cc !

Apart from the very small detail about the legacy icons within `cc-header-app` / `cc-tile-instances`, it's all good from my point of view.

Thanks again for all this work, can't wait for it to get merge :rocket: !

pdesoyres-cc reviewed (approved):
Well done @roberttran-cc . So much tiny things done in this PR :clap:

hsablonniere commented:
@roberttran-cc Here's my final review ;-)

Congrats for all this work (along with the other contributors)

## blocking

* discuss: some icons feel too thin and not that easy to understand/grasp
  * the tick used in orga member, input clipboard success...
  * the phone icon in orga header
  * the green redirection arrow for TCP redirs
* issue: in orga members, the icons in circle buttons in waiting mode aren't faded properly
* issue: why is the "i" icon smaller in tile-metrics compared to the other tiles?

## Not blocking

* note: the icon for "stopped" status is the only one to use remix so it won't be the same as in the console but I cannot see any difference, so OK ;-)
* warning: don't forget to delete the following unsued icons once the PR is rebased on the new pricing components
  * arrows-left.svg
  * arrows-right.svg
  * delete.svg
  * down.svg
  * plus.svg
  * storage.svg
  * sum.svg
  * up.svg
  * user.svg
* note: the cc-toggle still uses an image URL, but is it a problem?
  * we'll revisit this later but for now I guess it's OK
  * FYI the image URL is used as is in the cc-assistant (and the component stories)

roberttran-cc commented:
@hsablonniere thanks for your feedback!

**_"discuss: some icons feel too thin and not that easy to understand/grasp"_**
I'll start a Slack thread to decide the course of action.

**_issue: in orga members, the icons in circle buttons in waiting mode aren't faded properly_**
Nice catch! Will fix.

**_issue: why is the "i" icon smaller in tile-metrics compared to the other tiles?_**
It's my bad, as the size differs on master as well. Will fix.

**_note: the cc-toggle still uses an image URL, but is it a problem?_**
True: I think this component should behave like `cc-button` and should allow image and icon as well. I'll create an issue to track this.

hsablonniere reviewed (approved):
> "discuss: some icons feel too thin and not that easy to understand/grasp"
> I'll start a Slack thread to decide the course of action.

:+1: OK for using Remix, we'll revisit in the future if necessary

> issue: in orga members, the icons in circle buttons in waiting mode aren't faded properly
> Nice catch! Will fix.

:heavy_check_mark: Fixed, thanks

> issue: why is the "i" icon smaller in tile-metrics compared to the other tiles?
> It's my bad, as the size differs on master as well. Will fix.

:heavy_check_mark: Fixed, thanks

> note: the cc-toggle still uses an image URL, but is it a problem?
> True: I think this component should behave like cc-button and should allow image and icon as well. I'll create an issue to track this.

:+1: Agreed, let's do this!

:clap: GG for this work @roberttran-cc, LGTM!!
